### PR TITLE
select hangup cause NORMAL_CLEARING when computing ACD

### PR DESCRIPTION
**Pull request by phildeb**

### CODE REVIEW

File: mod_prometheus.rs [hunk 1]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
 
     // Channel hangup complete
     id = freeswitchrs::event_bind("mod_prometheus", fsr::event_types::CHANNEL_HANGUP_COMPLETE, None, |e| {
-        if let Some(last_bridge_hup_cause) = e.header("variable_last_bridge_hangup_cause") {
-            if last_bridge_hup_cause == "NORMAL_CLEARING" {
+
+        if let Some(callerIdName) = e.header("Caller-Caller-ID-Name"){
+            if let Some(uniqueId )= e.header("Unique-ID") {
```
Comment (grumvalski) [inline mod_prometheus.rs:321]: better to use the variable_sip_call_id (or maybe extract the 2?)  to be able to relate to actual calls and traces.

File: mod_prometheus.rs [hunk 2]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
+
+            if hupCause == "NORMAL_CLEARING" {  // NORMAL_CLEARING or ORIGINATOR_CANCEL or NO_USER_RESPONSE
                 if let Some(billsecvar) = e.header("variable_billsec") {
                     let parsed_time = billsecvar.parse::<i64>();
                     if parsed_time.is_ok() {
                         let bill_seconds = parsed_time.unwrap() ;
+                        
                         if let Some(direction) = e.header("Call-Direction") {
                             if direction == "outbound" {
                                 COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().increment_by(bill_seconds as f64);
```
Comment (grumvalski) [inline mod_prometheus.rs:342]: why is f64? shouldn't it be u64?

File: mod_prometheus.rs [hunk 3]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
                                 COUNTERS[FSCounter::SessionsOutboundCallHangupComplete].lock().unwrap().increment();
-                                let acd = COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().value() /
-                                                                COUNTERS[FSCounter::SessionsOutboundCallHangupComplete].lock().unwrap().value();
+
+                                let totalSeconds: f64 = COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().value();
```
Comment (grumvalski) [inline mod_prometheus.rs:345]: same as above for f64

File: mod_prometheus.rs [hunk 4]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
-                                let acd = COUNTERS[FSCounter::SessionsInboundCallDurationTotal].lock().unwrap().value() /
-                                                COUNTERS[FSCounter::SessionsInboundCallHangupComplete].lock().unwrap().value();
+
+                                let totalSeconds: f64 = COUNTERS[FSCounter::SessionsInboundCallDurationTotal].lock().unwrap().value();
+                                let totalHup: f64 = COUNTERS[FSCounter::SessionsInboundCallHangupComplete].lock().unwrap().value();
+                                let acd = totalSeconds / totalHup;
+                                
                                 GAUGES[FSGauge::SessionsInboundACD].lock().unwrap().set(acd);
+
+                                fslog!(INFO, "Inbound bill:{:#?} sec. totalHup:{:#?} total:{:#?} sec. acd:{:#?}\n", bill_seconds, totalHup, totalSeconds, acd );
```
Comment (grumvalski) [inline mod_prometheus.rs:362]: same as above.

File: mod_prometheus.rs [hunk 5]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
-                                let acd = COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().value() /
-                                                                COUNTERS[FSCounter::SessionsOutboundCallHangupComplete].lock().unwrap().value();
+
+                                let totalSeconds: f64 = COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().value();
+                                let totalHup: f64 = COUNTERS[FSCounter::SessionsOutboundCallHangupComplete].lock().unwrap().value();
+                                let acd = totalSeconds / totalHup;
+                                                                
                                 GAUGES[FSGauge::SessionsOutboundACD].lock().unwrap().set(acd);
+
+                                fslog!(INFO, "Outbound bill:{:#?} sec. totalHup:{:#?} total:{:#?} sec. acd:{:#?}\n", bill_seconds, totalHup, totalSeconds, acd );
```
Comment (grumvalski) [inline mod_prometheus.rs:351]: add the call-id to the log.

File: mod_prometheus.rs [hunk 6]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
+                                let acd = totalSeconds / totalHup;
+                                
                                 GAUGES[FSGauge::SessionsInboundACD].lock().unwrap().set(acd);
+
+                                fslog!(INFO, "Inbound bill:{:#?} sec. totalHup:{:#?} total:{:#?} sec. acd:{:#?}\n", bill_seconds, totalHup, totalSeconds, acd );
                             }
                         }
                     } else {
-                        fslog!(ERROR, "Received CHANNEL_HANGUP_COMPLETE with parsed_time error in variable_billsec information\n");
+                        fslog!(ERROR, "CHANNEL_HANGUP_COMPLETE error parsing variable_billsec header\n");
```
Comment (grumvalski) [inline mod_prometheus.rs:366]: add call-id.

File: mod_prometheus.rs [hunk 7]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
                     }
-                } else {
-                    fslog!(ERROR, "Received CHANNEL_HANGUP_COMPLETE with no variable_billsec information\n");
-                }
+                }else {
+                    fslog!(ERROR, "CHANNEL_HANGUP_COMPLETE without variable_billsec header\n");
```
Comment (grumvalski) [inline mod_prometheus.rs:369]: as above.

File: mod_prometheus.rs [hunk 8]
```diff
@@ -316,36 +316,61 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
+                } 
+            }else {
+                fslog!(ERROR, "CHANNEL_HANGUP_COMPLETE without Hangup-Cause header\n");
```
Comment (grumvalski) [inline mod_prometheus.rs:372]: as above.

File: mod_prometheus.rs [hunk 9]
```diff
@@ -316,36 +316,68 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
-                                GAUGES[FSGauge::SessionsOutboundACD].lock().unwrap().set(acd);
-                            } else if direction == "inbound" {
-                                COUNTERS[FSCounter::SessionsInboundCallDurationTotal].lock().unwrap().increment_by(bill_seconds as f64);
-                                COUNTERS[FSCounter::SessionsInboundCallHangupComplete].lock().unwrap().increment();
-                                let acd = COUNTERS[FSCounter::SessionsInboundCallDurationTotal].lock().unwrap().value() /
-                                                COUNTERS[FSCounter::SessionsInboundCallHangupComplete].lock().unwrap().value();
-                                GAUGES[FSGauge::SessionsInboundACD].lock().unwrap().set(acd);
-                            }
+                        if direction == "outbound" {
+                            COUNTERS[FSCounter::SessionsOutboundCallDurationTotal].lock().unwrap().increment_by(bill_seconds as f64);
```
Comment (grumvalski) [inline mod_prometheus.rs:345]: shouldn't it be u64 as well?

File: mod_prometheus.rs [hunk 10]
```diff
@@ -316,36 +316,68 @@ fn prometheus_load(mod_int: &ModInterface) -> Status {
...
+                            let acd_out = totalSeconds / totalHup;
+
+                            GAUGES[FSGauge::SessionsOutboundACD].lock().unwrap().set(acd_out as f64);
+
+                            fslog!(INFO, "callid:{:#?} uniqueId:{:#?} {:#?} bill:{:#?} sec. totalHup:{:#?} total:{:#?} sec. acd:{:#?} \n",
+                                callid, uniqueId, direction, bill_seconds, totalHup, totalSeconds, acd_out);
+
+                        } else if direction == "inbound" {
+
+                            COUNTERS[FSCounter::SessionsInboundCallDurationTotal].lock().unwrap().increment_by(bill_seconds as f64);
```
Comment (grumvalski) [inline mod_prometheus.rs:359]: shouldn't it be u64 as well?

Comment (phildeb) [reply mod_prometheus.rs:359]: no, this is how the function has been designed by Moises Sylva

File: mod_prometheus.rs (hunk 9 above)
Comment (phildeb) [reply mod_prometheus.rs:345]: no, this is how the function has been designed by Moises Sylva

**Pull request merged**